prefect_bitbucket/repository.py

```diff
@@ -0,0 +1,184 @@
+    )
+
+    branch_bitbucket_block.save(name="my-bitbucket-block")
+
+
+    # private BitBucket repository
+    private_bitbucket_block = BitBucketRepository(
+        repository="https://bitbucket.com/my-project/my-repository.git",
+        access_token="MY_BITBUCKET_PERSONAL_ACCESS_TOKEN"
+    )
+
+    private_bitbucket_block.save(name="my-private-bitbucket-block")
+```
+
+"""
+
+import io
+from distutils.dir_util import copy_tree
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Optional, Tuple, Union
+from urllib.parse import urlparse, urlunparse
+
+from prefect.exceptions import InvalidRepositoryURLError
+from prefect.filesystems import ReadableDeploymentStorage
+from prefect.utilities.asyncutils import sync_compatible
+from prefect.utilities.processutils import run_process
+from pydantic import Field, validator
+
+from prefect_bitbucket.credentials import BitBucketCredentials
+
+
+class BitBucketRepository(ReadableDeploymentStorage):
+    """
+    Interact with files stored in BitBucket repositories.
+
+    An accessible installation of git is required for this block to function
+    properly.
+    """
+
+    _block_type_name = "BitBucket Repository"
+    _logo_url = "https://images.ctfassets.net/gm98wzqotmnx/27LMR24ewTSDW238Lks1vH/34c5028659f4007528feadc8db8cecbd/500px-Bitbucket-blue-logomark-only.svg.png?h=250"  # noqa
+    _description = "Interact with files stored in BitBucket repositories."
+
+    repository: str = Field(
+        default=...,
+        description="The URL of a BitBucket repository to read from in HTTPS format",
+    )
+    reference: Optional[str] = Field(
+        default=None,
+        description="An optional reference to pin to; can be a branch or tag.",
+    )
+    credentials: Optional[BitBucketCredentials] = Field(
+        default=None,
+        description="An optional BitBucket Credentials block for authenticating with "
+        "private BitBucket repos.",
+    )
+
+    @validator("credentials")
+    def _ensure_credentials_go_with_https(cls, v: str, values: dict) -> str:
+        """Ensure that credentials are not provided with 'SSH' formatted BitBucket URLs.
+
+        Note: validates `access_token` specifically so that it only fires when private
+        repositories are used.
+        """
+        if v is not None:
+            if urlparse(values["repository"]).scheme != "https":
+                raise InvalidRepositoryURLError(
+                    (
+                        "Credentials can only be used with BitBucket repositories "
+                        "using the 'HTTPS' format. You must either remove the "
+                        "credential if you wish to use the 'SSH' format and are not "
+                        "using a private repository, or you must change the repository "
+                        "URL to the 'HTTPS' format."
+                    )
+                )
+
+        return v
```

Comment:
Can we add a test where `v is None`?

Reply:
v shouldn't be None since we are only running it if a credential is provided. What would be ensuring in that case, since there would be no credentials to compare the scheme with?

Reply:
Sorry, I meant a unit test that passes through this validator when `credentials=None`